Add ItemScope::visibility_of

edwin0cheng opened the pull request:
~This PR implements `HasVisibility` for various constructs and change `Definition::search_scope` to use `Visibility` directly instead of depends on ad-hoc string parsing.~

This PR added `visibility_of`  in `ItemScope` and `Module` and use it directly directly instead of depends on ad-hoc string parsing.

And also add a FIXME to indicate that there is a bug which do not search child-submodules in other files recursively in `Definition::search_scope`.

I will submit another PR to fix that bug after this is merged.

cc @flodiebold

Veetaha commented on crates/ra_hir/src/code_model.rs:
```diff
@@ -368,6 +375,14 @@ impl Struct {
+    fn visibility(&self, db: &dyn HirDatabase) -> Visibility {
+        let struct_data = db.struct_data(self.id);
+        let visibility = &struct_data.visibility;
+        visibility.resolve(db.upcast(), &self.id.resolver(db.upcast()))
```
Maybe @matklad can clarify. Do we by-design always accept `HirDatabase` even if we only need `DefDatabase` in `ra_hir`?

edwin0cheng replied:
In this case, we need `HirDatabase`, it is because the `trait` is called by same crate which should only exposed `HirDatabase` interface. I doubt that it is salsa limitation such that it has to be using an inheritance structure instead of composition to represent the hierarchy.   But feel free to correct me if i miss understand.

matklad commented:
One reason that we don't have visibility directly on declaraions is that visibility is not really a property of a declaration per se. Rather, it is a property of an "import" which brings the entyty into scope. So I wonder if a better API here would be to add `ItemScope::visibility_of(def)`?

flodiebold commented:
Yeah, I intentionally didn't add these visibilities.

edwin0cheng commented:
> Yeah, I intentionally didn't add these visibilities. Getting them from the scope might be a better idea.

Interesting. I never thought about that. I just removed all of these `HasVisibility` implementations and added `visibility_of` instead.

matklad commented:
bors r+

Nice! I've wanted to fix this for ages, thanks @edwin0cheng !

kjeremy commented on crates/ra_hir_def/src/item_scope.rs:
```diff
@@ -68,6 +68,12 @@ impl ItemScope {
         self.impls.iter().copied()
     }
 
+    pub fn visbility_of(&self, def: ModuleDefId) -> Option<Visibility> {
```
`visibility_of`?

edwin0cheng replied:
Oops.
